code/core/src/highlight/useHighlights.ts

```diff
@@ -582,37 +593,11 @@ export const useHighlights = ({
 
   channel.on(HIGHLIGHT, addHighlight);
   channel.on(REMOVE_HIGHLIGHT, removeHighlight);
-  channel.on(RESET_HIGHLIGHT, clearHighlights);
-  channel.on(STORY_CHANGED, clearHighlights);
+  channel.on(RESET_HIGHLIGHT, resetState);
   channel.on(SCROLL_INTO_VIEW, scrollIntoView);
-
-  const teardown = () => {
-    clearTimeout(removeTimeout);
-
-    document.body.removeEventListener('mousemove', onMouseMove);
-
-    channel.off(HIGHLIGHT, addHighlight);
-    channel.off(RESET_HIGHLIGHT, clearHighlights);
-    channel.off(STORY_CHANGED, clearHighlights);
-    channel.off(SCROLL_INTO_VIEW, scrollIntoView);
-
-    highlights.teardown();
-    elements.teardown();
-    boxes.teardown();
-    targets.teardown();
-    clickCoords.teardown();
-    hoverCoords.teardown();
-    hovered.teardown();
-    focused.teardown();
-    selected.teardown();
-
-    styleElementByHighlight.forEach((style) => style.remove());
-    boxElementByTargetElement.forEach((box) => box.remove());
-    document.getElementById(menuId)?.remove();
-    document.getElementById(rootId)?.remove();
-  };
-
-  (globalThis as any).__STORYBOOK_HIGHLIGHT_TEARDOWN = teardown;
-
-  return teardown;
+  channel.on(STORY_RENDER_PHASE_CHANGED, ({ newPhase }: { newPhase: string }) => {
+    if (newPhase === 'loading') {
+      resetState();
+    }
+  });
```

Comment:
out of scope for this PR, but I wanted to raise it anyway:
I pretty sure all of this code would become unnecessary if we gave this addon some more attention and turned it into a decorator.

There's quite a bit of code here, for the sake of synchronizing with storybook's rendering lifecycle, which is something that's near automatic in decorators as they are a part of said lifecycle.

Reply:
It would probably make sense for this to be a (global) decorator rather than a stand-alone script in the preview. Definitely not something to address right now though.